Proof HTTP cache page (#1156)
ezsystems/developer-documentation: docs/guide/http_cache.md

                        s_maxage: 20
```

> ### Setting time-to-live value for Page blocks

Author: I'd rather leave the hyphens here, as this is a compound modifier

> ### Setting time-to-live value for Page blocks

Reviewer: Yup, but like three lines earlier the hyphens are not there. So either add them in both places or remove here. WDYS?

> ### Setting time-to-live value for Page blocks

Author: True. "time to live" as a noun, "time-to-live" as a modifier. Overkill?

    FASTLY_KEY="token"
    ```

>     ##### Configuring Fastly on Platform.sh

Reviewer 2: Is there a way we can deal with link breakage on things like this? (and other header changes)

Author: Only on page level, not on heading level AFAIK

Reviewer 2: ok

3\. Add `Vary` in your custom controller or content view controller:

```php
$response->setVary('X-User-Preference-Hash');

// If you _also_ need to vary on eZ permissions, instead use:
//$response->setVary(['X-Context-User-Hash', 'X-User-Preference-Hash']);
```

## Content-aware HTTP cache

HTTP cache in eZ Platform is aware of which content or entity it is connected to.
This awareness is accomplished by means of cache tagging. All supported reverse proxies are content-aware.

!!! note "Tag header is stripped in production for security reasons"

    For security reasons this header, and other internal cache headers,
    are stripped from output in production by the reverse proxy (in VCL for Varnish and Fastly).

### Understanding cache tags

Understanding tags is the key to making the most of `ezplatform-http-cache`.

Tags form a secondary set of keys assigned to every cache item, on top of the "primary key" which is the URI.

> Like an index in a database, a tag is typically used for anything relevant, that represents the given cache item.

Reviewer: ```suggestion
Like an index in a database, a tag is typically used for anything relevant that represents the given cache item.
```

Tags are used for cache invalidation.

For example, the system tags every article response, and when the article Content Type is updated,
it tells Varnish that all articles should be considered stale
and be updated in the background when someone requests them.

Current content tags (and when the system purges on them):

- Content: `c<content-id>` - Purged on all smaller or larger changes to Content (including its metadata, fields and locations).
- Content Type: `ct<content-type-id>` - Used when the Content Type changes, affecting Content of its type.
- Location: `l<location-id>` - Used for clearing all cache relevant for a given Location.
- Parent Location: `pl<[parent-]location-id>` - Used for clearing all children of a Location (`pl<location-id>`), or all siblings (`pl<parent-location-id>`).
- Path: `p<location-id>` - For operations that change the tree itself, like move, remove, etc.
- Relation: `r<content-id>` - Only purged on when content updates are severe enough to also affect reverse relations.
- Relation location: `rl<location-id>` - Same as relation, but by Location ID.

!!! note "Automatic repository prefixing of cache tags"

>     As eZ Platform support multi-repository (multi-database) setups that can have overlapping IDs,

Reviewer: ```suggestion
    As eZ Platform supports multi-repository (multi-database) setups that can have overlapping IDs,
```

    the shared HTTP cache systems need to distinguish tags relevant to the different content repositories.

    This is why in multi-repository setup you can see cache tags such as `1p2`.

>     In this example `1` represents the index among configured repositories, meaning the second repository in the system.

Reviewer: ```suggestion
    In this example, `1` represents the index among configured repositories, meaning the second repository in the system.
```

>     Tags are not prefixed for default repository (index "0").

Reviewer: ```suggestion
    Tags are not prefixed for the default repository (index "0").
```

#### Troubleshooting - Cache header too long errors

> In case of complex content, for instance Pages with many blocks, or RichText with a lot of embeds/links,

Reviewer: ```suggestion
In case of complex content, for example, Pages with many blocks, or RichText with a lot of embeds/links,
```

you can encounter problems with too long cache header on responses.
Because of this, necessary cache entries may not be tagged properly.

> You may also see `502 Headers too long` errors, and webserver refusing to serve the page.

Reviewer: ```suggestion
You may also see `502 Headers too long` errors, and a web server refusing to serve the page.
```

###### A. Configure  allowing larger headers

Varnish configuration:

- [http_resp_hdr_len](https://varnish-cache.org/docs/6.0/reference/varnishd.html#http-resp-hdr-len) (default 8k, change to i.e. 32k)
- [http_max_hdr](https://varnish-cache.org/docs/6.0/reference/varnishd.html#http-max-hdr) (default 64, change to i.e. 128)
- [http_resp_size](https://varnish-cache.org/docs/6.0/reference/varnishd.html#http-resp-size) (default 23k, change to i.e. 96k)
- [workspace_backend](https://varnish-cache.org/docs/6.0/reference/varnishd.html#workspace-backend) (default 64k, change to i.e. 128k)

If you need to see these long headers in `varnishlog`, adapt the [vsl_reclen](https://varnish-cache.org/docs/6.0/reference/varnishd.html#vsl-reclen) setting.

Nginx has a default limit of 4k/8k when buffering responses:

- For [PHP-FPM](https://www.php.net/manual/en/install.fpm.php) setup using proxy module, configure [proxy_buffer_size](https://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_buffer_size)
- For FastCGI setup using fastcgi module, configure [fastcgi_buffer_size](https://nginx.org/en/docs/http/ngx_http_fastcgi_module.html#fastcgi_buffer_size)

> Fastly has a `Surrogate-Key` header limit of 16 kB, this cannot be changed.

Reviewer: ```suggestion
Fastly has a `Surrogate-Key` header limit of 16 kB, which cannot be changed.
```

> Apache has a [hard](https://github.com/apache/httpd/blob/5f32ea94af5f1e7ea68d6fca58f0ac2478cc18c5/server/util_script.c#L495) [coded](https://github.com/apache/httpd/blob/7e2d26eac309b2d79e467ef586526c10e0f226f8/include/httpd.h#L299-L303) limit of 8 kB, so if you face this issue consider using Nginx instead.

Reviewer: ```suggestion
Apache has a [hard](https://github.com/apache/httpd/blob/5f32ea94af5f1e7ea68d6fca58f0ac2478cc18c5/server/util_script.c#L495) [coded](https://github.com/apache/httpd/blob/7e2d26eac309b2d79e467ef586526c10e0f226f8/include/httpd.h#L299-L303) limit of 8 kB, so if you face this issue, consider using Nginx instead.
```
Two external links linked from each word of the "hard coded" phrase. Could it possibly be split somehow?

###### B. Limit tags header output by system

1\. For inline rendering just displaying the content name, image attribute, and/or link, it would be enough to:

- Look into how many inline (non ESI) render calls for content rendering you are doing, and see if you can organize it differently.
- Consider inlining the views not used elsewhere in the given template and [tagging the response in Twig](#response-tagging-in-twig) with "relation" tags.

>     - Optionally, you can set reduced cache TTL for the given view in order to reduce risk of stale cache on subtree operations affecting the inlined content.

Reviewer: ```suggestion
    - Optionally, you can set a reduced cache TTL for the given view in order to reduce a risk of stale cache on subtree operations affecting the inlined content.
```

    ezplatform.http_cache.tags.header_reduced_ttl: 7200
```

### Response tagging done with content view

> For content views response tagging is done automatically, and cache system outputs headers like the following:

Reviewer: ```suggestion
For content views, response tagging is done automatically, and a cache system outputs headers that are similar to the following one:
```

```
HTTP/1.1 200 OK
Cache-Control: public, max-age=86400
xkey: ez-all c1 ct1 l2 pl1 p1 p2
```

If the given content has several Locations, you can see several `l<location-id>` and `p<location-id>` tags in the response.

!!! note "How response tagging for ContentView is done internally"

    In `ezplatform-http-cache` there is a dedicated response listener `HttpCacheResponseSubscriber` that checks if:
    
    - the response has attribute `view`
    - the view implements `eZ\Publish\Core\MVC\Symfony\View\CachableView`
    - cache is not disabled on the individual view

    If that checks out, the response is adapted with the following:
    
    - `ResponseCacheConfigurator` applies SiteAccess settings for enabled/disabled cache and default TTL
    - `DispatcherTagger` dispatches the built-in ResponseTaggers which generate the tags as described above.

>     Further information can be found in [`ezplatform-http-cache` docs.](https://github.com/ezsystems/ezplatform-http-cache/blob/master/docs/response_taggers.md)

Reviewer: ```suggestion
    Further information can be found in [`ezplatform-http-cache` documentation.](https://github.com/ezsystems/ezplatform-http-cache/blob/master/docs/response_taggers.md)
```

### Response tagging in controllers

For tagging needs in controllers, there are several options, here presented in recommended order:

1\. Reusing `DispatcherTagger` to pick correct tags.

> Examples for tagging everything needed for content using the autowirable `ResponseTagger` interface:

Reviewer: Does it mean "Examples for using the autowirable `ResponseTagger` interface to tag everything that is needed for the content:" ?

> Examples for tagging everything needed for content using the autowirable `ResponseTagger` interface:

Author: ```suggestion
Examples for using the autowirable `ResponseTagger` interface to tag everything that is needed for the content:
```

$tagHandler->addRelationTags([33, 44]);
```

> 3\. Manually add tags yourself using low-level FOS `TagHandler`.

Reviewer: ```suggestion
3\. Add tags manually by using low-level FOS `TagHandler`.
```

In PHP, FOSHttpCache exposes the `fos_http_cache.http.symfony_response_tagger` service which enables you to add tags to a response.

The following example adds minimal tags for when ID 33 and 34 are rendered in ESI,

> but parent response needs these tags to get refresh if they are deleted:

Reviewer: "to get refresh" - "to get refreshed"?

> but parent response needs these tags to get refresh if they are deleted:

Author: ```suggestion
but parent response needs these tags to get refreshed if they are deleted:
```

4\. Use deprecated `X-Location-Id` header.

For custom or eZ controllers (e.g. REST) still using `X-Location-Id`, `XLocationIdResponseSubscriber` handles translating
this header to tags. It supports singular and comma-separated Location ID value(s):

```php
/** @var \Symfony\Component\HttpFoundation\Response $response */
$response->headers->set('X-Location-Id', 123);

// Alternatively using several Location ID values
$response->headers->set('X-Location-Id', '123,212,42');
```

!!! caution "X-Location-Id use is deprecated"

>     `X-Location-Id` is deprecated and will be removed in future.

Reviewer: ```suggestion
    `X-Location-Id` is deprecated and will be removed in the future.
```

>     For rendering content it is advised to refactor to use Content View,

Reviewer: ```suggestion
    For rendering content, it is advised to refactor to use Content View,
```

Also, the whole sentence seems like missing verbs to connect the nouns, esp. the second part.

### Tag purging

#### Default purge tagging

`ezplatform-http-cache` uses Repository API event subscribers to listen to events emitted on Repository operations,
and depending on the operation triggers expiry on a specific tag or set of tags.

> For example on the move Location event the following tags are purged:

Reviewer: ```suggestion
For example, on the move Location event, the following tags are purged:
```

`ezplatform-http-cache` uses Repository API event subscribers to listen to events emitted on Repository operations,
and depending on the operation triggers expiry on a specific tag or set of tags.

> For example, on the move Location event, the following tags are purged:

Reviewer: If it is a distinct event name, would it make sense to make "move Location" standa out, like, put it in quotes?

];
```

All event subscribers can be found in `ezplatform-http-cache/src/EventSubscriber/CachePurge`.

#### Custom purging from code

While the system purges tags whenever API is used to change data, you may need to purge directly from code.

> For that you can use the built-in purge client:

Reviewer: ```suggestion
You can use the built-in purge client for that:
```
